src/mongodbTypes.ts

```diff
@@ -213,21 +225,21 @@ export interface PaprUpdateFilter<TSchema> {
         $type: 'date' | 'timestamp';
       }
   >;
-  $inc?: OnlyFieldsOfType<TSchema, NumericType | undefined>;
-  $min?: PaprMatchKeysAndValues<TSchema>;
-  $max?: PaprMatchKeysAndValues<TSchema>;
-  $mul?: OnlyFieldsOfType<TSchema, NumericType | undefined>;
+  $inc?: OnlyFieldsOfType<DeepReadonly<TSchema>, NumericType | undefined>;
+  $min?: PaprMatchKeysAndValues<DeepReadonly<TSchema>>;
+  $max?: PaprMatchKeysAndValues<DeepReadonly<TSchema>>;
+  $mul?: OnlyFieldsOfType<DeepReadonly<TSchema>, NumericType | undefined>;
   $rename?: Record<string, string>;
-  $set?: PaprMatchKeysAndValues<TSchema>;
-  $setOnInsert?: PaprMatchKeysAndValues<TSchema>;
-  $unset?: OnlyFieldsOfType<TSchema, any, '' | 1 | true>;
-  $addToSet?: SetFields<TSchema>;
-  $pop?: OnlyFieldsOfType<TSchema, readonly any[], -1 | 1>;
-  $pull?: PullOperator<TSchema>;
-  $push?: PushOperator<TSchema>;
-  $pullAll?: PullAllOperator<TSchema>;
+  $set?: PaprMatchKeysAndValues<DeepReadonly<TSchema>>;
+  $setOnInsert?: PaprMatchKeysAndValues<DeepReadonly<TSchema>>;
+  $unset?: OnlyFieldsOfType<DeepReadonly<TSchema>, any, '' | 1 | true>;
+  $addToSet?: SetFields<DeepReadonly<TSchema>>;
+  $pop?: OnlyFieldsOfType<DeepReadonly<TSchema>, readonly any[], -1 | 1>;
+  $pull?: PullOperator<DeepReadonly<TSchema>>;
+  $push?: PushOperator<DeepReadonly<TSchema>>;
+  $pullAll?: PullAllOperator<DeepReadonly<TSchema>>;
   $bit?: OnlyFieldsOfType<
-    TSchema,
+    DeepReadonly<TSchema>,
```

Comment:
All this wrapping seems not DRY at all. Can we optimize this code somehow?

Reply:
I didn't see a simple way to do so, though I feel the duplication is pretty well concentrated, living only here where the major tributaries of the update methods feed into the `PaprUpdateFilter` defintion, and nowhere else.